File: socketd/src/main/java/org/noear/socketd/client/ClientBase.java
package org.noear.socketd.client;

import org.noear.socketd.core.HeartbeatHandler;
import org.noear.socketd.core.Listener;
import org.noear.socketd.core.ChannelAssistant;
import org.noear.socketd.core.Processor;
import org.noear.socketd.core.impl.ProcessorDefault;

import java.util.function.Consumer;

/**
 * 客户端基类
 *
 * @author noear
 * @since 2.0
 */
public abstract class ClientBase<T extends ChannelAssistant> implements Client {
    protected Processor processor = new ProcessorDefault();
    protected HeartbeatHandler heartbeatHandler;

    private final ClientConfig config;
    private final T assistant;

    public ClientBase(ClientConfig clientConfig, T assistant) {
        this.config = clientConfig;
        this.assistant = assistant;
    }

    /**
     * 配置
     */
    public ClientConfig config() {
        return config;
    }

    /**
     * 通道助理
     */
    public T assistant() {
        return assistant;
    }

    /**
     * 处理器
     */
    public Processor processor() {
        return processor;
    }

    @Override
    public Client heartbeatHandler(HeartbeatHandler handler) {
        if (handler != null) {
            this.heartbeatHandler = handler;
        }

        return this;
    }

    public HeartbeatHandler heartbeatHandler() {
        return heartbeatHandler;
    }

    public long heartbeatInterval() {
        return config.getHeartbeatInterval();
    }

    @Override
    public Client config(Consumer<ClientConfig> consumer) {
        consumer.accept(config);
        return this;
    }

    @Override
    public Client listen(Listener listener) {
        processor.setListener(listener);
        return this;
    }
}

File: go/socketd/transport/client/client_base.go
package client

import (
	"time"

	"socketd/transport/core"
	"socketd/transport/core/impl"
)

type ClientBase[T core.ChannelAssistant[U], U any] struct {
	processor        core.Processor
	heartbeatHandler core.HeartbeatHandler
	config           *Config
	assistant        T
	connect          Connector
}

func NewClientBase[T core.ChannelAssistant[U], U any](config *Config, assistant T) *ClientBase[T, U] {
	var cb = new(ClientBase[T, U])
	cb.processor = impl.NewProcessor()
	cb.config = config
	cb.assistant = assistant
	return cb
}

func (cb *ClientBase[T, U]) Connect(connect Connector) {
	cb.connect = connect
}

func (cb *ClientBase[T, U]) GetAssistant() T {
	return cb.assistant
}

func (cb *ClientBase[T, U]) GetHeartbeatHandler() core.HeartbeatHandler {
	return cb.heartbeatHandler
}

func (cb *ClientBase[T, U]) GetHeartbeatInterval() time.Duration {
	return cb.config.GetHeartbeatInterval()
}

func (cb *ClientBase[T, U]) GetConfig() *Config {
	return cb.config
}

func (cb *ClientBase[T, U]) GetProcessor() core.Processor {
	return cb.processor
}

func (cb *ClientBase[T, U]) HeartbeatHandler(handler core.HeartbeatHandler) Client {
	cb.heartbeatHandler = handler
	return cb
}

//func (cb *ClientBase[T, U]) Config(configHandler ClientConfigHandler) Client {
//	//TODO implement me
//	panic("implement me")
//}

func (cb *ClientBase[T, U]) Listen(listener core.Listener) Client {
	cb.processor.SetListener(listener)
	return cb
}

func (cb *ClientBase[T, U]) Open() (Session, error) {
	conn, err := cb.connect.Connect()
	if err != nil {
		return nil, err
	}
	clientChannel := NewClientChannel(conn, cb.connect)
	clientChannel.SetHandshake(conn.GetHandshake())
	session := impl.NewSessionDefault(clientChannel)
	return session, nil
}
